### PR TITLE
member add/delete, snapshot save/restore functions implemented using clientv3

### PR DESCRIPTION
**Pull request by retroflexer**
This library provides utility functions for disaster recovery (DR). These are golang port of the existing shell script [openshift-recovery-tools.sh](https://github.com/hexfusion/machine-config-operator/blob/2143b18b87dd64c64bc166dd6138b9074d25f933/templates/master/00-master/_base/files/usr-local-bin-openshift-recovery-tools-sh.yaml)

### CODE REVIEW

File: etcdutils.go [hunk 1]
```diff
@@ -77,7 +77,7 @@ users:
 `
 	t := template.Must(template.New("config").Parse(configTemplate))
 	if err = t.Execute(os.Stdout, params); err != nil {
-		log.Fatal(err)
+		fmt.Fprintf(os.Stderr, "Error executing template %v\n", err)
```
Comment (mfojtik) [inline etcdutils.go]: the `log.Fatal()` will `os.Exit(1)`, are you aware that `fmt.Fprintf` does not do that?

File: etcdutils.go [hunk 2]
```diff
@@ -152,28 +148,29 @@ func RemoveCerts(etcdStaticResourceDir string) {
...
-func RestoreSnapshot() {
-	// NOT needed, direct etcdctl call
-}
+func PatchManifest(manifestFilePath, old, new string) {
 
-func PatchManifest() {
-}
+	read, err := ioutil.ReadFile(manifestFilePath)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "%s:read error %s", manifestFilePath, err)
```
Comment (mfojtik) [inline etcdutils.go]: nit: use %v for printing errors

File: etcdutils.go [hunk 3]
```diff
@@ -152,28 +148,29 @@ func RemoveCerts(etcdStaticResourceDir string) {
...
+	}
+	fmt.Println(manifestFilePath)
 
-func EtcdMemberAdd() {
-	// Direct call to etcdctl
+	newContents := strings.Replace(string(read), old, new, -1)
+
+	err = ioutil.WriteFile(manifestFilePath, []byte(newContents), 0)
```
Comment (mfojtik) [inline etcdutils.go]: style nit: this is the case when you write `if err:= ...; err !=nil {}` ... you don't need the `err` below

File: etcdutils_clientv3.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+	"time"
+
+	"github.com/coreos/etcd/pkg/fileutil"
+	"go.etcd.io/etcd/clientv3"
+	"go.etcd.io/etcd/clientv3/snapshot"
+	"go.etcd.io/etcd/embed"
+	"go.uber.org/zap"
+)
+
+func SaveSnapshot(ctx context.Context, cfg clientv3.Config, dbPath string) error {
```
Comment (mfojtik) [inline etcdutils_clientv3.go:18]: every public func should have comments

File: etcdutils_clientv3.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+	if len(cfg.Endpoints) != 1 {
+		return fmt.Errorf("snapshot must be requested to one selected node, not multiple %v", cfg.Endpoints)
```
Comment (mfojtik) [inline etcdutils_clientv3.go]: nit: use `%#v` to print slices

File: etcdutils_clientv3.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+	partpath := dbPath + ".part"
+	defer os.RemoveAll(partpath)
+
+	var f *os.File
+	f, err = os.OpenFile(partpath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, fileutil.PrivateFileMode)
+	if err != nil {
+		return fmt.Errorf("could not open %s (%v)", partpath, err)
+	}
+	fmt.Println("created temporary db file", partpath)
```
Comment (mfojtik) [inline etcdutils_clientv3.go]: is this debugging leftover?

File: etcdutils_clientv3.go [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+	now := time.Now()
+	var rd io.ReadCloser
+	rd, err = cli.Snapshot(ctx)
+	if err != nil {
+		return err
+	}
+	fmt.Println("fetching snapshot from:", cfg.Endpoints[0])
```
Comment (mfojtik) [inline etcdutils_clientv3.go]: nit: debugging? if not, i would maybe use `log` package that gives you timestamps?

File: etcdutils_clientv3.go [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+	if err = os.Rename(partpath, dbPath); err != nil {
+		return fmt.Errorf("could not rename %s to %s (%v)", partpath, dbPath, err)
+	}
+	fmt.Println("saved snapshot to path", dbPath)
+	return nil
+}
+
+func RestoreSnapshot(ctx context.Context, cfg embed.Config, peerURLs []string, dbPath string) error {
+	sp := snapshot.NewV3(zap.NewExample())
+	if err := sp.Restore(snapshot.RestoreConfig{
```
Comment (mfojtik) [inline etcdutils_clientv3.go]: just `return sp.Restore(...)`

File: etcdutils_clientv3.go [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+	if err != nil {
+		return err
+	}
+	defer cli.Close()
+
+	mresp, err := cli.MemberAdd(context.Background(), peerURLs)
+	if err != nil {
+		return err
+	}
+	fmt.Println("added member.PeerURLs:", mresp.Member.PeerURLs)
```
Comment (mfojtik) [inline etcdutils_clientv3.go]: use `log`

File: etcdutils_clientv3.go [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+	return nil
+}
+
+func EtcdMemberRemove(ctx context.Context, cfg clientv3.Config, memberName string) error {
+	var id uint64 = 0
```
Comment (mfojtik) [inline etcdutils_clientv3.go]: move this above the for loop (line 108)

File: etcdutils_clientv3.go [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+			id = m.ID
+			break
+		}
+	}
+
+	if id == 0 {
+		return fmt.Errorf("member not found to remove")
+	}
+
+	_, err = cli.MemberRemove(context.Background(), id)
```
Comment (mfojtik) [inline etcdutils_clientv3.go]: pass the `ctx` ?

File: etcdutils_clientv3.go [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+
+func EtcdMemberRemove(ctx context.Context, cfg clientv3.Config, memberName string) error {
+	var id uint64 = 0
+	cli, err := clientv3.New(cfg)
+	if err != nil {
+		return err
+	}
+	defer cli.Close()
+
+	resp, err := cli.MemberList(context.Background())
```
Comment (mfojtik) [inline etcdutils_clientv3.go]: pass the `ctx` ?

File: etcdutils_clientv3.go [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+}
+
+func EtcdMemberAdd(ctx context.Context, cfg clientv3.Config, peerURLs []string) error {
+	cli, err := clientv3.New(cfg)
+	if err != nil {
+		return err
+	}
+	defer cli.Close()
+
+	mresp, err := cli.MemberAdd(context.Background(), peerURLs)
```
Comment (mfojtik) [inline etcdutils_clientv3.go]: pass the `ctx`

File: etcdutils.go (hunk 1 above)
Comment (retroflexer) [reply etcdutils.go]: That's exactly why I removed log.Fatal and added error, and return below the line.

File: etcdutils.go (hunk 2 above)
Comment (retroflexer) [reply etcdutils.go]: Yeah. I used %v elsewhere, but been inconsistent.

**Pull request merged**